x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -601,84 +610,66 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
-                    // Making the following call slow down. Maybe we can think something smarter.
-                    Number n = context.doc().getByKey(delegateFieldMapper.name()).numericValue();
-                    if (n.intValue() < 0) {
-                        throw new IllegalArgumentException(
-                            "Aggregate metric [" + metric.name() + "] of field [" + mappedFieldType.name() + "] cannot be a negative number"
-                        );
-                    }
-                } else if (Metric.max == metric) {
-                    max = context.doc().getByKey(delegateFieldMapper.name()).numericValue();
-                } else if (Metric.min == metric) {
-                    min = context.doc().getByKey(delegateFieldMapper.name()).numericValue();
+                    metricsParsed.put(metric, delegateFieldMapper.value(context.parser()));
+                } catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException("failed to parse [" + metric.name() + "] sub field: " + e.getMessage(), e);
                 }
-                metricsParsed.add(metric);
                 token = subParser.nextToken();
             }
+
             // check max value must bigger then min value
+            Number min = metricsParsed.get(Metric.min);
+            Number max = metricsParsed.get(Metric.max);
             if (max != null && min != null && max.doubleValue() < min.doubleValue()) {
                 throw new IllegalArgumentException(
                     "Aggregate metric field [" + mappedFieldType.name() + "] max value cannot be smaller than min value"
                 );
             }
-
-            // Check if all required metrics have been parsed.
-            if (metricsParsed.containsAll(metrics) == false) {
+            Number valueCount = metricsParsed.get(Metric.value_count);
+            if (valueCount != null && valueCount.intValue() < 0) {
                 throw new IllegalArgumentException(
-                    "Aggregate metric field [" + mappedFieldType.name() + "] must contain all metrics " + metrics.toString()
+                    "Aggregate metric [" + Metric.value_count + "] of field [" + mappedFieldType.name() + "] cannot be a negative number"
                 );
             }
+
+            // Check if all required metrics have been parsed.
+            for (Metric m : metrics) {
+                if (metricsParsed.containsKey(m) == false) {
+                    throw new IllegalArgumentException(
+                        "Aggregate metric field [" + mappedFieldType.name() + "] must contain all metrics " + metrics
+                    );
+                }
+            }
+            // Check that there aren't any duplicates already parsed
+            for (Metric m : metricsParsed.keySet()) {
+                NumberFieldMapper delegateFieldMapper = metricFieldMappers.get(m);
+                if (context.doc().getByKey(delegateFieldMapper.fieldType().name()) != null) {
```

Comment:
Method `Document.onlyAddKey()` was supposed to check for duplicate fields in the doc. I think this method is called from `NumferFieldMapper.indexValue()`. In this case the duplicates check is performed twice.

Reply:
Yeah. I suppose I don't need to check this. It's just passing the right exception message back. I *think* don't need to check this because a duplicate will fail with the first key every time - because we don't index unless all keys are present. That's what I did above.

Reply:
I think this method deserves another visit because I see us constantly doing a test like this one to get a better error message. I'd like to smooth it a bit in a follow up if that's ok
.